planner/cascades/transformation_rules.go

```diff
@@ -1282,6 +1283,64 @@ func (r *PushTopNDownTiKVSingleGather) OnTransform(old *memo.ExprIter) (newExprs
 	return []*memo.GroupExpr{finalTopNExpr}, true, false, nil
 }
 
+// MergeAdjacentTopN merge adjacent TopN.
+type MergeAdjacentTopN struct {
+	baseRule
+}
+
+// NewRuleMergeAdjacentTopN creates a new Transformation MergeAdjacentTopN.
+// The pattern of this rule is `TopN->TopN->X`.
+func NewRuleMergeAdjacentTopN() Transformation {
+	rule := &MergeAdjacentTopN{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandTopN, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentTopN) Match(expr *memo.ExprIter) bool {
+	topN := expr.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	child := expr.Children[0].GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	if len(topN.ByItems) != len(child.ByItems) {
+		return false
+	}
+	for i := 0; i < len(topN.ByItems); i++ {
+		if !topN.ByItems[i].Equal(topN.SCtx(), child.ByItems[i]) {
+			return false
+		}
```

Comment:
Can we consider to merge these two `TopN` as well: `TOpN(a, b) -> TopN(a)`? Which is, the sort columns of parent `TopN` is a prefix of child `TopN`.

Reply:
Good catch. In my opinion, it may cause some mistakes. For example, when the data are ```(1, 3), (1, 2), (1, 1)``` and the operators are ```top1(a, b) -> top2(a)```. Your idea gets the operator ```top1(a, b)``` and the answer will be ```(1, 1)```. But the true answer should be ```(1, 2)```.  @zz-jason

Reply:
Since the root operator is `TopN(a)` which only requires to order by `a`, so I think any of `(1, 3)`, `(1, 2)`, or `(1, 1)` is a valid result when `1` is the smallest value in the table and `N` equals to 1.